### PR TITLE
chore(sessions): break out sessions manager code and add unit tests

### PR DESCRIPTION
**Pull request by emilyrohrbough**
Apart of https://github.com/cypress-io/cypress/issues/20977

This refactor is prepping for updating changes with cy.session to use the new command group API and adding test coverage for GA support.

This PR breaks out the cy.session session manager functions into its own file and adds some missing unit tests -- there are a few that are still need to be added that I will update at a later time.

### User facing changelog

n/a

### PR Tasks

- [x] Have tests been added/updated?
- [n/a] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [n/a] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cy/commands/sessions/manager.ts [hunk 1]
```diff
@@ -0,0 +1,295 @@
+import _ from 'lodash'
```
Comment (mschile) [inline packages/driver/src/cy/commands/sessions/manager.ts:1]: Did you change any logic in this file?

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 2]
```diff
@@ -0,0 +1,354 @@
+const SessionsManager = require('../../../../src/cy/commands/sessions/manager').default
+const $Cypress = require('../../../../src/cypress').default
+
+describe('src/cy/commands/sessions/manager.ts', () => {
+  let CypressInstance
+  let baseUrl
+
+  beforeEach(function () {
+  // @ts-ignore
```
Comment (chrisbreiding) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:9]:
```suggestion
    // @ts-ignore
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    const sessionsManager = new SessionsManager(CypressInstance, () => {})
+
+    expect(sessionsManager).to.haveOwnProperty('cy')
+    expect(sessionsManager).to.haveOwnProperty('Cypress')
+    expect(sessionsManager).to.haveOwnProperty('currentTestRegisteredSessions')
+    expect(sessionsManager.currentTestRegisteredSessions).to.be.instanceOf(Map)
+  })
+
+  describe('.setActiveSession()', () => {
+    it('adds session when non were previously added', () => {
```
Comment (chrisbreiding) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:24]:
```suggestion
    it('adds session when none were previously added', () => {
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+      sessionsManager.setActiveSession(activeSession)
+      const calls = cySpy.getCalls()
+
+      expect(cySpy).to.be.calledTwice
+      expect(calls[0].args[1]).to.be.undefined
+      expect(calls[1].args[1]).to.haveOwnProperty('session_1')
+    })
+
+    it('adds session when other sessions had previously added', () => {
```
Comment (chrisbreiding) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:45]:
```suggestion
    it('adds session when other sessions were previously added', () => {
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+      const origins = await sessionsManager.mapOrigins(['*', 'https://other.com'])
+
+      expect(origins).to.deep.eq(['https://example.com', baseUrl, 'http://foobar.com', 'https://other.com'])
+      expect(sessionsSpy).to.be.calledOnce
+    })
+  })
+
+  describe('._setStorageOnOrigins()', () => {})
+
+  it('.getAllHtmlOrigins()', async () => {
```
Comment (chrisbreiding) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:209]: Should this be a `describe` with an `it` that describes the method's behavior?

File: packages/driver/src/cy/commands/sessions/manager.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    const clearedSessions: ActiveSessions = _.mapValues(curSessions, (v) => ({ ...v, hydrated: false }))
+
+    this.cy.state('activeSessions', clearedSessions)
+  }
+
+  mapOrigins = async (origins: string | Array<string>): Promise<Array<string>> => {
+    const getOrigins = this.Cypress.Promise.map(
+      ([] as string[]).concat(origins), async (v) => {
+        if (v === '*') {
+          return await this.getAllHtmlOrigins()
```
Comment (emilyrohrbough) [inline packages/driver/src/cy/commands/sessions/manager.ts:48]: Updated to use existing function that handled this logic
```diff
- return _.keys(await Cypress.backend('get:rendered:html:origins')).concat([currentOrigin])
+ return await this.getAllHtmlOrigins()
```

File: packages/driver/src/cy/commands/sessions/manager.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+      if (currentOriginIndex !== -1) {
+        origins.splice(currentOriginIndex, 1)
+        const currentOriginStorage = getCurrentOriginStorage()
+
+        pushValue(currentOrigin, currentOriginStorage)
+      }
+
+      if (_.isEmpty(origins)) {
+        return results
```
Comment (emilyrohrbough) [inline packages/driver/src/cy/commands/sessions/manager.ts:238]: Updates
```diff
- return getResults()
- return results
```

File: packages/driver/src/cy/commands/sessions/manager.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        _.remove(origins, (v) => v.startsWith('http:'))
+      }
+
+      const postMessageResults = await getPostMessageLocalStorage(specWindow, origins)
+
+      postMessageResults.forEach((val) => {
+        pushValue(val[0], val[1])
+      })
+
+      return results
```
Comment (emilyrohrbough) [inline packages/driver/src/cy/commands/sessions/manager.ts:251]: Updates
```diff
- return getResults()
- return results
```

File: packages/driver/src/cy/commands/sessions/manager.ts (hunk 1 above)
Comment (emilyrohrbough) [reply packages/driver/src/cy/commands/sessions/manager.ts:1]: Made comments on the changes.

**Review by chrisbreiding — APPROVED** [review]

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+      const sessionsSpy = cy.stub(sessionsManager, 'getAllHtmlOrigins').resolves(allOrigins)
+
+      const origins = await sessionsManager.mapOrigins(['*', 'https://other.com'])
+
+      expect(origins).to.deep.eq(['https://example.com', baseUrl, 'http://foobar.com', 'https://other.com'])
+      expect(sessionsSpy).to.be.calledOnce
+    })
+  })
+
+  describe('._setStorageOnOrigins()', () => {})
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:207]: I assume you meant for this describe to wrap the test below?

Comment (emilyrohrbough) [reply packages/driver/cypress/integration/commands/sessions/manager_spec.ts:207]: no, I missed adding a `// TO DO` comment.

Comment (emilyrohrbough) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:207]:
```suggestion
  // TODO
  describe('._setStorageOnOrigins()', () => {})
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts (hunk 5 above)
Comment (emilyrohrbough) [reply packages/driver/cypress/integration/commands/sessions/manager_spec.ts:209]: I don't think it matters. I can change it if you feel strongly.

**Review by mjhenkes — APPROVED** [review]

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+
+      await sessionsManager.sessions.clearCurrentSessionData()
+
+      expect(clearStorageSpy).to.be.calledOnce
+      expect(clearCookiesSpy).to.be.calledOnce
+      expect(window.localStorage).of.have.lengthOf(0)
+      expect(window.sessionStorage).of.have.lengthOf(0)
+    })
+
+    // TODO
```
Comment (AtofStryker) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:282]: I wonder if it is good to add the colon here to help some of the vscode plugins pick up the comment?
```suggestion
    // TODO:
```

File: packages/driver/src/cy/commands/sessions/manager.ts (hunk 6 above)
Comment (AtofStryker) [reply packages/driver/src/cy/commands/sessions/manager.ts:48]: 🙏

**Review by AtofStryker — APPROVED** [review]

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+      const sessionsSpy = cy.stub(sessionsManager, 'getAllHtmlOrigins').resolves(allOrigins)
+
+      const origins = await sessionsManager.mapOrigins(['*', 'https://other.com'])
+
+      expect(origins).to.deep.eq(['https://example.com', baseUrl, 'http://foobar.com', 'https://other.com'])
+      expect(sessionsSpy).to.be.calledOnce
+    })
+  })
+
+  // TODO
```
Comment (emilyrohrbough) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:207]:
```suggestion
  // TODO:
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+      const cypressSpy = cy.stub(CypressInstance, 'backend').callThrough().withArgs('get:session')
+
+      const sessionsManager = new SessionsManager(CypressInstance, () => {})
+
+      sessionsManager.sessions.getSession('session_1')
+
+      expect(cypressSpy).to.be.calledOnceWith('get:session', 'session_1')
+    })
+
+    // TODO
```
Comment (emilyrohrbough) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:346]:
```suggestion
    // TODO:
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    describe('sessions.getStorage', () => {})
+
+    // TODO
```
Comment (emilyrohrbough) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:349]:
```suggestion
    // TODO:
```

File: packages/driver/cypress/integration/commands/sessions/manager_spec.ts [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+    describe('sessions.clearStorage', () => {})
+
+    // TODO
```
Comment (emilyrohrbough) [inline packages/driver/cypress/integration/commands/sessions/manager_spec.ts:352]:
```suggestion
    // TODO:
```

**Pull request merged**